Only show menu items from same parent menu

joshuatf opened the pull request:
Fixes #5718

Fixes an issue where menu items could have the same parent, but different menu IDs and still be shown together.

### Detailed test instructions:

1. Register a menu item in the plugins menu, but using a parent ID from the main menu:
```php

		\Automattic\WooCommerce\Admin\Features\Navigation\Menu::add_plugin_item(
			array(
				'id'     => 'wrong-menu-item',
				'parent' => 'analytics',
				'title'  => 'I should not be shown',
				'url'    => 'https://www.youtube.com/watch?v=dQw4w9WgXcQ',
			)
		);
```
2. Navigate to that category.
3. Make sure the item is not shown.
4. Make sure no other regressions occur with items showing across various menu categories.

joshuatf committed:
Only render nav items belonging to parent menu ID

joshuatf committed:
Set item menu ID for items automatically

joelclimbsthings commented on client/navigation/components/container/index.js:
```diff
@@ -87,6 +87,26 @@ const Container = ( { menuItems } ) => {
+
+		const isTopLevel = category.id === 'woocommerce';
+
+		if ( ! isTopLevel && items[ 0 ].menuId !== category.menuId ) {
```
Looking at the code, I think we can expect that this would never be called where `items[ 0 ]` had the same menu ID, but `items[ 1 ]` did not?

joshuatf replied:
Right, since they're already categorized all items should have the same menu ID, though I think there's probably a better way to handle this.

joelclimbsthings commented on client/navigation/components/container/index.js:
```diff
@@ -128,44 +148,18 @@ const Container = ( { menuItems } ) => {
-									</NavigationGroup>
+								{ renderNavigationGroup(
+									category,
+									pluginItems,
```
So it looks like this will often be called with `0` items for most categories, which returns null as the first line of the function. Works well, although I wonder if we could refactor this to make the code clearer at some point? 

I don't think this has any real performance consequences so I'm not too worried in the context of this PR.

joshuatf replied:
I don't think there's any measurable performance impact, but what would you like to see to make it clearer?

We could remove the `return null` logic and instead wrap each of the groups with an item check.   This isn't very DRY, but if you feel it makes it clearer then I'm not opposed.
```js
{ primaryItems.length && renderNavigationGroup(
	category,
	primaryItems
) }
{ pluginItems.length && renderNavigationGroup(
	category,
	pluginItems,
	__( 'Extensions', 'woocommerce-admin' )
) }
{ secondaryItems.length && renderNavigationGroup(
	category,
	secondaryItems
) }
```

joelclimbsthings replied:
Looks like this is the approach after the recent changes. Not a big deal at all, but I think this is definitely a bit clearer coming at it with no foreknowledge of what's happening. I like it :+1:

joelclimbsthings reviewed (commented):
Hey @joshuatf , looks and works great! Glad to see we were sync'd up with our thinking on this. 

I left a couple comments I was curious about, but likely don't require changes at this point. Just curious what you think.

Also, I tested this using your test instructions and all was well (the item didn't show up), ~~although I actually noticed that the problem didn't show up with the current state of `main` for me either. So I wasn't able to reproduce for :shrug:~~

Edit: Scratch that last comment, it was just because I was listing the parent as `analytics` instead of `woocommerce-analytics`

joshuatf committed:
Use categories map to filter items early

joshuatf committed:
Convert categorized menus to objects

joshuatf committed:
Prevent bypassing menu via existing parent

joshuatf committed:
Fix plugins name

joshuatf commented:
Thanks for the review @joelclimbsthings - I refactored some of the existing code in https://github.com/woocommerce/woocommerce-admin/pull/5731/commits/51c1b73314ea830f2e6d4fabb9743a18a09a27d9 so that we can filter categorized items ahead of time.

Also found another loophole to move an item to another menu; fixed in https://github.com/woocommerce/woocommerce-admin/pull/5731/commits/4dd5b45e2989eebe1e5e92f7d9207219dc7d9982

joelclimbsthings reviewed (approved):
Good catch on that last loophole :1st_place_medal: 

And I like the refactored approach as well. :shipit:
